src/zenml/metadata/lazy_load.py

```diff
@@ -0,0 +1,66 @@
+#  Copyright (c) ZenML GmbH 2024. All Rights Reserved.
+#
+#  Licensed under the Apache License, Version 2.0 (the "License");
+#  you may not use this file except in compliance with the License.
+#  You may obtain a copy of the License at:
+#
+#       https://www.apache.org/licenses/LICENSE-2.0
+#
+#  Unless required by applicable law or agreed to in writing, software
+#  distributed under the License is distributed on an "AS IS" BASIS,
+#  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
+#  or implied. See the License for the specific language governing
+#  permissions and limitations under the License.
+"""Run Metadata Lazy Loader definition."""
+
+from typing import TYPE_CHECKING, Optional
+from uuid import uuid4
+
+if TYPE_CHECKING:
+    from zenml.model.model_version import ModelVersion
+    from zenml.models import RunMetadataResponse
+
+
+class RunMetadataLazyGetter:
+    """Run Metadata Lazy Getter helper class.
+
+    It serves the purpose to feed back to the user the metadata
+    lazy loader wrapper for any given key, if called inside a pipeline
+    design time context.
+    """
+
+    def __init__(
+        self,
+        _lazy_load_model_version: "ModelVersion",
+        _lazy_load_artifact_name: Optional[str],
+        _lazy_load_artifact_version: Optional[str],
+    ):
+        """Initialize a RunMetadataLazyGetter.
+
+        Args:
+            _lazy_load_model_version: The model version.
+            _lazy_load_artifact_name: The artifact name.
+            _lazy_load_artifact_version: The artifact version.
+        """
+        self._lazy_load_model_version = _lazy_load_model_version
+        self._lazy_load_artifact_name = _lazy_load_artifact_name
+        self._lazy_load_artifact_version = _lazy_load_artifact_version
+
+    def __getitem__(self, key: str) -> "RunMetadataResponse":
+        """Get the metadata for the given key.
+
+        Args:
+            key: The metadata key.
+
+        Returns:
+            The metadata lazy loader wrapper for the given key.
+        """
+        from zenml.models import RunMetadataResponse
+
+        return RunMetadataResponse(
+            id=uuid4(),
```

Comment:
As you already pointed out in the PR description, not a huge fan of this. I don't think we can easily remove that from the `BaseResponse` class (for our actual responses, we do always want the ID to be in there). There's also a problem when users in their pipeline function try to access any properties of the lazy responses, they will get random values or strange error messages.

At least the error messages we could prevent by maybe having a separate subclass for it, but not convinced this is a good idea..
```python
class LazyArtifactVersionResponse(ArtifactVersionResponse):
     _lazy_load_name: Optional[str] = None
    _lazy_load_version: Optional[str] = None
    _lazy_load_model_version: Optional["ModelVersion"] = None

    # We should probably do this for all properties, otherwise users will run into strange error message when they 
    # try to access any of those on a lazy object
    @property
    def id(self) -> UUID:
        raise RuntimeError("ID can't be accessed for lazy AVR")

    ...
```

Reply:
Actually instead of overwriting all properties, we probably only need to overwrite the `get_body()` and `get_metadata()` functions to raise errors which describe the issue.

Reply:
Super, thanks!